index.js

```diff
@@ -1,174 +1,255 @@
+    return
+  }
+  const waiter = this._pendingQueue.shift()
+  if (this._idle.length) {
+    const idleItem = this._idle.pop()
+    clearTimeout(idleItem.timeoutId)
+    return waiter(null, idleItem.client)
+  }
+  if (!this._isFull()) {
+    return this.connect(waiter)
+  }
+  throw new Error('unexpected condition')
 }
 
-Pool.prototype._promiseNoCallback = function (callback, executor) {
-  return callback
-    ? executor()
-    : new this.Promise(executor)
+Pool.prototype._promisify = function (callback) {
+  if (callback) {
+    return { callback: callback, result: undefined }
+  }
+  let reject
+  let resolve
+  const cb = function (err, client) {
+    err ? reject(err) : resolve(client)
+  }
+  const result = new this.Promise(function (res, rej) {
+    resolve = res
+    reject = rej
+  })
+  return { callback: cb, result: result }
 }
 
-Pool.prototype._destroy = function (client) {
-  if (client._destroying) return
-  client._destroying = true
+Pool.prototype._remove = function (client) {
+  this._idle = this._idle.filter(item => item.client !== client)
+  this._clients = this._clients.filter(c => c !== client)
   client.end()
+  this.emit('remove', client)
 }
 
-Pool.prototype._create = function (cb) {
-  this.log('connecting new client')
-  var client = new this.Client(this.options)
+class IdleItem {
+  constructor (client, timeoutId) {
+    this.client = client
+    this.timeoutId = timeoutId
+  }
+}
 
-  client.on('error', function (e) {
-    this.log('connected client error:', e)
-    this.pool.destroy(client)
-    e.client = client
-    this.emit('error', e, client)
-  }.bind(this))
+function release (client, err) {
+  client.release = function () { throw new Error('called release twice') }
+  if (err) {
+    this._remove(client)
+    this._pulseQueue()
+    return
+  }
 
-  client.connect(function (err) {
-    if (err) {
-      this.log('client connection error:', err)
-      cb(err, null)
-    } else {
-      this.log('client connected')
-      this.emit('connect', client)
-      cb(null, client)
-    }
-  }.bind(this))
+  // idle timeout
+  let tid
+  if (this.options.idleTimeoutMillis) {
+    tid = setTimeout(() => {
+      this.log('remove idle client')
+      this._remove(client)
+    }, this.idleTimeoutMillis)
+  }
+
+  if (this.ending) {
+    this._remove(client)
+  } else {
+    this._idle.push(new IdleItem(client, tid))
+  }
+  this._pulseQueue()
 }
 
 Pool.prototype.connect = function (cb) {
-  return this._promiseNoCallback(cb, function (resolve, reject) {
-    this.log('acquire client begin')
-    this.pool.acquire(function (err, client) {
+  if (this.ending) {
+    const err = new Error('Cannot use a pool after calling end on the pool')
+    return cb ? cb(err) : this.Promise.reject(err)
+  }
+  if (this._clients.length >= this.options.max || this._idle.length) {
+    const response = this._promisify(cb)
+    const result = response.result
+    cb = response.callback
+    this._pendingQueue.push((err, client) => {
```

Comment:
proposal: instead of  handling this in a closure -- just store the waiter and handle this when invoking it during dequeue/shift